### DISCUSSION TITLE
Open from URL?

### DISCUSSION THREAD

**Discussion by mattsmith321** [UglyToad/PdfPig#370] [Q&A] [answered]
I am trying to use PdfPig in an NUnit project with Selenium to verify the existence of content in the PDF. Is there a way to open/load the PDF from the URL that the webdriver currently has?

Still looking around for examples but everything references local files. 

Thanks!

**Comment by EliotJones** [comment, marked as answer]
You'd want to download the file before passing to the library, you can do this using `HttpClient` on NET 4.5+ or `WebClient` for earlier versions. You might be able to pass the content stream directly to PdfPig but I  don't believe the stream supports seeking so you should download the bytes first, e.g. https://stackoverflow.com/a/54475013

```
var client = new HttpClient();
var response = await client.GetAsync("https://my-url.com/pdf");
if (!response.IsSuccessStatusCode) { /* handle the error */ }
var bytes = await response.Content.ReadAsBytesAsync();
using var pdf = PdfDocument.Open(bytes);
```

I haven't checked this in the IDE so some of the methods may be named differently.

**Reply by mattsmith321** [reply]
Thanks for the response. Still working through it on my end. There are aspects to it that are making it challenging such as authentication. Still trying to make it work though.

**Reply by EliotJones** [reply]
This is more a Selenium specific thing but generally you will need to retrieve the authentication data, e.g. cookies or localStorage, from the browser and attach them to the request in the way your server expects, e.g. Bearer token. Use the network tab of the browser to figure out what you need to provide in the request to get the file.

https://www.selenium.dev/documentation/support_packages/working_with_cookies/#get-all-cookies

**Reply by mattsmith321** [reply]
Yes, lots of constraints trying to make it work within the context of Selenium. 

Particularly challenging because all I have is an <embed> object with no src to download the file from:

`<body><embed name="1A88405F004E107C9721749DC2F2653A" src="about:blank" type="application/pdf" internalid="1A88405F004E107C9721749DC2F2653A"></body>`

I ended up with the following which seems to be working. I know WebClient is deprecated but I was struggling with HttpClient originally.

`        StringBuilder cookies = new StringBuilder();

        string encoded = Convert.ToBase64String(Encoding.ASCII.GetBytes(
            string.Format("{0}:{1}", Username, Password)));

        foreach (OpenQA.Selenium.Cookie cookie in driver.Manage().Cookies.AllCookies)
        {
            string value = cookie.Name + "=" + cookie.Value;
            if (cookies.Length == 0)
            {
                cookies.Append(value);
            }
            else
            {
                cookies.Append(";").Append(value);
            }
        }

        using (WebClient client = new WebClient())
        {
            client.Headers.Add("Authorization", "Basic " + encoded);
            client.Headers.Add("Cookie", cookies.ToString());
            client.DownloadFile(driver.Url, filename);
        }

        using (PdfDocument document = PdfDocument.Open(filename))
        {
            // Page number starts from 1, not 0.
            Page page = document.GetPage(1);

            Assert.True(page.Text.Contains(expectedText), string.Format("PDF does not contain expected text '{0}'", expectedText));
        }

        File.Delete(filename);
`

Thanks for the PdfPig library. It worked great!
